### PR TITLE
Refactored Navigator into BasicNavigator and Navigator

### PR DESCRIPTION
**Pull request by gspencergoog**
**Note:** _at the moment, this is a strawman: I'm not sure of several of the decisions here, so I'm mainly looking for comment on the approach (at least from @Hixie, who encouraged me to see if I could find a solution).  I haven't updated most of the doc strings yet, or written a new test for BasicNavigator (although all the old tests pass), and there are probably formatting issues._

The goal here was to factor out the `Navigator`'s named route functionality so that we can later support different kinds of route-building from names, such as from URLs, which at least two platforms use for inter-app communication.

I did this by splitting `Navigator` into `BasicNavigator` and `Navigator` (which now extends `BasicNavigator`), and by factoring out a base class for their states (still called `NavigatorState`, unfortunately).  The `Navigator` would have liked to be renamed `NamedNavigator`, but that was just too invasive of a change. I did change the `Route<T>` `navigator` field to be called `navigatorState`, which is much more accurate (I did leave the old `navigator` accessor, but deprecated it). The `BasicNavigator` has all the functionality of the old `Navigator`, except that it doesn't have anything that deals with named routes or the initial route as a string.

It is unfortunate that `NavigatorState` was originally exposed in the `Route<T>` class, because it means that the API for `NavigatorState` now needs to be supported for all navigators, even ones that don't support `pushNamed`, etc.  I've done this by making a base class out of `NavigatorState` that the new `_NavigatorState` and `_BasicNavigatorState` extend.  The base class now has to be a generic that takes a `BasicNavigator` subclass, which does cause some weak warnings in the analyzer about not specifying the generic type.  It works, but that's probably the most annoying part of this change.

I haven't looked too deeply into what changes this implies to the *App classes, but there could probably be some changes there.

### CODE REVIEW

File: packages/flutter/lib/src/widgets/navigator.dart [hunk 1]
```diff
@@ -30,9 +30,15 @@ import 'ticker_provider.dart';
 /// See [MaterialPageRoute] for a route that replaces the
 /// entire screen with a platform-adaptive transition.
 abstract class Route<T> {
-  /// The navigator that the route is in, if any.
-  NavigatorState get navigator => _navigator;
-  NavigatorState _navigator;
+  /// The navigator that the route is in, if any.  Deprecated, use navigatorState
+  /// getter instead.
+  @deprecated
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/navigator.dart]: when deprecating, do something like `@Deprecated('Use navigatorState instead')`

File: packages/flutter/lib/src/widgets/navigator.dart [hunk 2]
```diff
@@ -30,9 +30,15 @@ import 'ticker_provider.dart';
...
+  NavigatorState<BasicNavigator> get navigator => _navigatorState;
+
+  /// The [NavigatorState] of the [BasicNavigator] that the route is in,
+  /// if any.
+  NavigatorState<BasicNavigator> get navigatorState => _navigatorState;
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/navigator.dart]: why this change?

File: packages/flutter/lib/src/widgets/navigator.dart [hunk 3]
```diff
@@ -1334,7 +1288,9 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin {
       // If we're between frames (SchedulerPhase.idle) then absorb any
       // subsequent pointers from this frame. The absorbing flag will be
       // reset in the next frame, see build().
-      final RenderAbsorbPointer absorber = _overlayKey.currentContext?.ancestorRenderObjectOfType(const TypeMatcher<RenderAbsorbPointer>());
+      final RenderAbsorbPointer absorber = _overlayKey.currentContext
+          ?.ancestorRenderObjectOfType(
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/navigator.dart]: definitely don't line break before an elvis operator, that looks really weird :-)

File: packages/flutter/lib/src/widgets/navigator.dart [hunk 4]
```diff
@@ -1352,7 +1308,8 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin {
       onPointerUp: _handlePointerUpOrCancel,
       onPointerCancel: _handlePointerUpOrCancel,
       child: new AbsorbPointer(
-        absorbing: false, // it's mutated directly by _cancelActivePointers above
+        absorbing: false,
+        // it's mutated directly by _cancelActivePointers above
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/navigator.dart]: the line break makes it very unclear what the comment refers to

File: packages/flutter/lib/src/widgets/navigator.dart [hunk 5]
```diff
@@ -731,26 +664,111 @@ class Navigator extends StatefulWidget {
...
       }
       return true;
     });
-    return navigator;
+    return navigatorState;
   }
 
   @override
-  NavigatorState createState() => new NavigatorState();
+  _NavigatorState createState() => new _NavigatorState();
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/navigator.dart]: createState should always be last

**Comment by Hixie** [comment]
Usually we would prefer composition rather than inheritance.

It seems like BasicNavigator here still has named route functionality (and as you say, given the APIs we've committed to, I'm not sure we can ever change that).

A different approach might be to have a delegate that you can pass to the Navigator which is then in charge of splitting the initial route, saying which initial routes to push, and converting new route names to routes. Maybe onGenerateRoute and onUnknownRoute would be shorthands for creating a delegate with those methods; we do that in other places sometimes.

Fundamentally though I think we first need to decide what problem we're solving, i.e. list all the use cases and discuss which we're not handling today, before writing code. I don't feel I have a good enough handle on the problem space to evaluate the solutions.

File: packages/flutter/lib/src/widgets/navigator.dart (hunk 3 above)
Comment (gspencergoog) [inline packages/flutter/lib/src/widgets/navigator.dart]: Yikes!  Yeah, I didn't do that on purpose.  I think IntelliJ did that to "help" me when I pasted code.  I noticed it did a few other crazy things too (putting line breaks into ? : conditionals, etc.).  I'll have to look into changing that.

File: packages/flutter/lib/src/widgets/navigator.dart (hunk 1 above)
Comment (gspencergoog) [inline packages/flutter/lib/src/widgets/navigator.dart]: OK, sure.

File: packages/flutter/lib/src/widgets/navigator.dart (hunk 2 above)
Comment (gspencergoog) [inline packages/flutter/lib/src/widgets/navigator.dart]: Because there are still things that need access to the NavigatorState, unfortunately. It just bothered me that it was called "navigator", but what was being returned was _not_ a Navigator.

File: packages/flutter/lib/src/widgets/navigator.dart (hunk 5 above)
Comment (gspencergoog) [inline packages/flutter/lib/src/widgets/navigator.dart]: OK, noted.

File: packages/flutter/lib/src/widgets/navigator.dart (hunk 4 above)
Comment (gspencergoog) [inline packages/flutter/lib/src/widgets/navigator.dart]: Yeah, again, I didn't do that. I blame IntelliJ.

**Review by gspencergoog — COMMENTED** [review]
OK, all good points.  I'll see if I can write up a doc that contains as many use cases as I can think of, and maybe something will reveal itself.  The delegate route  (har har - no pun intended) seems promising.

I needed to try refactoring it to see where the weaknesses were, so this was a useful exercise, really. I'll probably just close this PR and the other one, since it doesn't seem like part of the the final solution (and if it turns out to be, I'll keep this branch anyhow).  I'm really not liking the generic base for NavigatorState, though: it makes it so client code has to change in order get rid of the warnings.

**Comment by gspencergoog** [comment]
> Maybe onGenerateRoute and onUnknownRoute would be shorthands for creating a delegate with those methods; we do that in other places sometimes.

Can you give me an example of one of the places?

**Comment by Hixie** [comment]
The ListView and GridView classes have various constructors that delegate their arguments to, well, delegates in various ways.
